File: code/mcp/server/typelab.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"os"
	"path/filepath"
)

type TypeLabInput struct {
	Preset                 string  `json:"preset"`
	BodySizeChangePercent  float64 `json:"body_size_change_percent"`
	TargetCapHeightDegrees float64 `json:"target_cap_height_degrees"`
	LinePitchChangePercent float64 `json:"line_pitch_change_percent"`
	ViewingDistanceInches  float64 `json:"viewing_distance_inches"`
}

type typeLabFace struct {
	AscentUnits    int `json:"ascent_units"`
	DescentUnits   int `json:"descent_units"`
	LineGapUnits   int `json:"line_gap_units"`
	CapHeightUnits int `json:"cap_height_units"`
	FBInkPixels    int `json:"fbink_pixels"`
}

func (device *SSHDevice) TypeLab(input TypeLabInput) (map[string]any, error) {
	if err := validatePresetID(input.Preset); err != nil {
		return nil, err
	}
	if input.BodySizeChangePercent < -40 || input.BodySizeChangePercent > 100 {
		return nil, errors.New("body_size_change_percent must be between -40 and 100")
	}
	if input.LinePitchChangePercent < -20 || input.LinePitchChangePercent > 100 {
		return nil, errors.New("line_pitch_change_percent must be between -20 and 100")
	}
	if input.TargetCapHeightDegrees != 0 && input.BodySizeChangePercent != 0 {
		return nil, errors.New("use target_cap_height_degrees or body_size_change_percent, not both")
	}

	path := filepath.Join(device.Root, "kindle", "reader", "presets", "mac", input.Preset+".json")
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read preset %s: %w", input.Preset, err)
	}
	var preset struct {
		Calibration struct {
			ViewingDistance float64 `json:"viewing_distance_inches"`
			PanelPPI        float64 `json:"panel_pixels_per_inch"`
		} `json:"calibration"`
		Fonts struct {
			Narrative typeLabFace `json:"narrative"`
		} `json:"fonts"`
	}
	if err := json.Unmarshal(raw, &preset); err != nil {
		return nil, fmt.Errorf("parse preset %s: %w", input.Preset, err)
	}
	face := preset.Fonts.Narrative
	if err := validateTypeLabFace(face); err != nil {
		return nil, fmt.Errorf("preset %s: %w", input.Preset, err)
	}
	distance := input.ViewingDistanceInches
	if distance == 0 {
		distance = preset.Calibration.ViewingDistance
	}
	if distance < 6 || distance > 60 {
		return nil, errors.New("viewing_distance_inches must be between 6 and 60")
	}
	if preset.Calibration.PanelPPI <= 0 {
		return nil, errors.New("preset panel PPI must be positive")
	}

	currentPitch, currentGap := lineGeometry(face, face.FBInkPixels)
	candidatePixels := int(math.Round(float64(face.FBInkPixels) * (1 + input.BodySizeChangePercent/100)))
	if input.TargetCapHeightDegrees != 0 {
		if input.TargetCapHeightDegrees < 0.15 || input.TargetCapHeightDegrees > 1.5 {
			return nil, errors.New("target_cap_height_degrees must be between 0.15 and 1.5")
		}
		candidatePixels = pixelsForAngle(input.TargetCapHeightDegrees, distance, preset.Calibration.PanelPPI, face)
	}
	if candidatePixels < 8 {
		return nil, errors.New("candidate body size is below 8 pixels")
	}

	targetPitch := int(math.Round(float64(currentPitch) * (1 + input.LinePitchChangePercent/100)))
	baseHeight := unspacedHeight(face, candidatePixels)
	requestedGap := targetPitch - baseHeight
	if requestedGap < 0 {
		requestedGap = 0
	}
	lineGapUnits := minimumLineGapUnits(requestedGap, candidatePixels, face)
	candidateFace := face
	candidateFace.LineGapUnits = lineGapUnits
	candidatePitch, candidateGap := lineGeometry(candidateFace, candidatePixels)
	angle := achievedTypeAngle(candidatePixels, distance, preset.Calibration.PanelPPI, face)

	return map[string]any{
		"ok":     true,
		"preset": input.Preset,
		"current": map[string]any{
			"body_fbink_pixels":     face.FBInkPixels,
			"line_gap_pixels":       currentGap,
			"baseline_pitch_pixels": currentPitch,
		},
		"candidate": map[string]any{
			"body_fbink_pixels":             candidatePixels,
			"line_gap_pixels":               candidateGap,
			"baseline_pitch_pixels":         candidatePitch,
			"achieved_cap_height_degrees":   angle,
			"viewing_distance_inches":       distance,
			"above_nasa_025_degree_minimum": angle >= 0.25,
		},
		"actual_change_percent": map[string]any{
			"body_size":  100 * (float64(candidatePixels)/float64(face.FBInkPixels) - 1),
			"line_pitch": 100 * (float64(candidatePitch)/float64(currentPitch) - 1),
		},
		"recipe": map[string]any{
			"narrative_fbink_pixels": candidatePixels,
			"hhea_line_gap_units":    lineGapUnits,
			"source_preset":          input.Preset,
		},
	}, nil
}

func validateTypeLabFace(face typeLabFace) error {
	if face.AscentUnits <= face.DescentUnits || face.CapHeightUnits <= 0 || face.FBInkPixels <= 0 {
		return errors.New("narrative font metrics are incomplete")
	}
	return nil
}

func unspacedHeight(face typeLabFace, pixels int) int {
	vertical := float64(face.AscentUnits - face.DescentUnits)
	baseline := int(math.Ceil(float64(pixels) * float64(face.AscentUnits) / vertical))
	descent := int(math.Ceil(float64(pixels) * float64(face.DescentUnits) / vertical))
	return baseline + int(math.Abs(float64(descent)))
}

func lineGeometry(face typeLabFace, pixels int) (pitch, gap int) {
	vertical := float64(face.AscentUnits - face.DescentUnits)
	gap = int(math.Ceil(float64(pixels) * float64(face.LineGapUnits) / vertical))
	return unspacedHeight(face, pixels) + gap, gap
}

func minimumLineGapUnits(gapPixels, pixels int, face typeLabFace) int {
	if gapPixels <= 0 {
		return 0
	}
	vertical := face.AscentUnits - face.DescentUnits
	// Smallest integer whose scaled ceiling reaches gapPixels.
	return int(math.Floor(float64((gapPixels-1)*vertical)/float64(pixels))) + 1
}

func pixelsForAngle(angle, distance, ppi float64, face typeLabFace) int {
	capPixels := 2 * distance * ppi * math.Tan(angle*math.Pi/360)
	vertical := float64(face.AscentUnits - face.DescentUnits)
	return int(math.Ceil(capPixels * vertical / float64(face.CapHeightUnits)))
}

func achievedTypeAngle(pixels int, distance, ppi float64, face typeLabFace) float64 {
	vertical := float64(face.AscentUnits - face.DescentUnits)
	capInches := (float64(pixels) * float64(face.CapHeightUnits) / vertical) / ppi
	return 2 * math.Atan(capInches/(2*distance)) * 180 / math.Pi
}

File: code/scripts/type-lab.py
#!/usr/bin/env python3
"""Calculate ReKindled type geometry without modifying a font or device."""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path
from typing import Any


def unspaced_height(face: dict[str, Any], pixels: int) -> int:
    vertical = face["ascent_units"] - face["descent_units"]
    baseline = math.ceil(pixels * face["ascent_units"] / vertical)
    descent = math.ceil(pixels * face["descent_units"] / vertical)
    return baseline + abs(descent)


def line_geometry(face: dict[str, Any], pixels: int) -> tuple[int, int]:
    vertical = face["ascent_units"] - face["descent_units"]
    gap = math.ceil(pixels * face.get("line_gap_units", 0) / vertical)
    return unspaced_height(face, pixels) + gap, gap


def minimum_line_gap_units(gap_pixels: int, pixels: int, face: dict[str, Any]) -> int:
    if gap_pixels <= 0:
        return 0
    vertical = face["ascent_units"] - face["descent_units"]
    return math.floor((gap_pixels - 1) * vertical / pixels) + 1


def pixels_for_angle(angle: float, distance: float, ppi: float, face: dict[str, Any]) -> int:
    cap_pixels = 2 * distance * ppi * math.tan(math.radians(angle) / 2)
    vertical = face["ascent_units"] - face["descent_units"]
    return math.ceil(cap_pixels * vertical / face["cap_height_units"])


def achieved_angle(pixels: int, distance: float, ppi: float, face: dict[str, Any]) -> float:
    vertical = face["ascent_units"] - face["descent_units"]
    cap_inches = (pixels * face["cap_height_units"] / vertical) / ppi
    return math.degrees(2 * math.atan(cap_inches / (2 * distance)))


def calculate(
    preset: dict[str, Any],
    *,
    body_change: float,
    target_angle: float | None,
    pitch_change: float,
    distance: float | None,
) -> dict[str, Any]:
    face = preset["fonts"]["narrative"]
    calibration = preset["calibration"]
    ppi = float(calibration["panel_pixels_per_inch"])
    distance = float(distance or calibration["viewing_distance_inches"])
    current_pixels = int(face["fbink_pixels"])
    current_pitch, current_gap = line_geometry(face, current_pixels)

    candidate_pixels = round(current_pixels * (1 + body_change / 100))
    if target_angle is not None:
        candidate_pixels = pixels_for_angle(target_angle, distance, ppi, face)
    if candidate_pixels < 8:
        raise ValueError("candidate body size is below 8 pixels")

    target_pitch = round(current_pitch * (1 + pitch_change / 100))
    base_height = unspaced_height(face, candidate_pixels)
    requested_gap = max(0, target_pitch - base_height)
    gap_units = minimum_line_gap_units(requested_gap, candidate_pixels, face)
    candidate_face = dict(face, line_gap_units=gap_units)
    candidate_pitch, candidate_gap = line_geometry(candidate_face, candidate_pixels)
    angle = achieved_angle(candidate_pixels, distance, ppi, face)

    return {
        "schema_version": 1,
        "source_preset": preset.get("id", "unknown"),
        "current": {
            "body_fbink_pixels": current_pixels,
            "line_gap_pixels": current_gap,
            "baseline_pitch_pixels": current_pitch,
        },
        "candidate": {
            "body_fbink_pixels": candidate_pixels,
            "line_gap_pixels": candidate_gap,
            "baseline_pitch_pixels": candidate_pitch,
            "achieved_cap_height_degrees": angle,
            "viewing_distance_inches": distance,
            "above_nasa_025_degree_minimum": angle >= 0.25,
        },
        "actual_change_percent": {
            "body_size": 100 * (candidate_pixels / current_pixels - 1),
            "line_pitch": 100 * (candidate_pitch / current_pitch - 1),
        },
        "recipe": {
            "narrative_fbink_pixels": candidate_pixels,
            "hhea_line_gap_units": gap_units,
            "source_preset": preset.get("id", "unknown"),
        },
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Turn physical typography dials into an FBInk/OpenType recipe."
    )
    parser.add_argument("preset", type=Path, help="source typography preset JSON")
    group = parser.add_mutually_exclusive_group()
    group.add_argument(
        "--body-change-percent",
        type=float,
        default=0,
        help="relative change to FBInk body pixels; -5 means five percent smaller",
    )
    group.add_argument(
        "--target-cap-height-degrees",
        type=float,
        help="choose body pixels from an angular cap-height target",
    )
    parser.add_argument(
        "--line-pitch-change-percent",
        type=float,
        default=0,
        help="change absolute baseline pitch relative to the source preset",
    )
    parser.add_argument("--viewing-distance-inches", type=float)
    parser.add_argument("--write-recipe", type=Path, help="also write the result as JSON")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    if not -40 <= args.body_change_percent <= 100:
        raise SystemExit("--body-change-percent must be between -40 and 100")
    if not -20 <= args.line_pitch_change_percent <= 100:
        raise SystemExit("--line-pitch-change-percent must be between -20 and 100")
    if args.target_cap_height_degrees is not None and not 0.15 <= args.target_cap_height_degrees <= 1.5:
        raise SystemExit("--target-cap-height-degrees must be between 0.15 and 1.5")
    if args.viewing_distance_inches is not None and not 6 <= args.viewing_distance_inches <= 60:
        raise SystemExit("--viewing-distance-inches must be between 6 and 60")

    preset = json.loads(args.preset.read_text(encoding="utf-8"))
    try:
        result = calculate(
            preset,
            body_change=args.body_change_percent,
            target_angle=args.target_cap_height_degrees,
            pitch_change=args.line_pitch_change_percent,
            distance=args.viewing_distance_inches,
        )
    except (KeyError, TypeError, ValueError, ZeroDivisionError) as exc:
        raise SystemExit(f"invalid preset or dial value: {exc}") from exc

    encoded = json.dumps(result, indent=2, sort_keys=True) + "\n"
    print(encoded, end="")
    if args.write_recipe:
        args.write_recipe.parent.mkdir(parents=True, exist_ok=True)
        args.write_recipe.write_text(encoded, encoding="utf-8")


if __name__ == "__main__":
    main()
